### PR TITLE
8332400: isspace argument should be a valid unsigned char

### PR DESCRIPTION
**Pull request by roberttoyonaga**
### Summary
This change ensures we don't get undefined behavior when calling[`isspace`](https://pubs.opengroup.org/onlinepubs/007904975/functions/isspace.html).  `isspace` accepts an `int` argument that "the application shall ensure is a character representable as an unsigned char or equal to the value of the macro EOF.". 

Previously, there was no checking of the values passed to `isspace`. I've replaced direct calls with a new wrapper `os::is_space` that performs a range check and prevents the possibility of undefined behavior from happening. For instances outside of Hotspot, I've added casts to `unsigned char`. 

**Testing**
- Added a new test in `test/hotspot/gtest/runtime/test_os.cpp` to check `os::is_space` is working correctly.
- tier1

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8332400: isspace argument should be a valid unsigned char`

### Issue
 * [JDK-8332400](https://bugs.openjdk.org/browse/JDK-8332400): isspace argument should be a valid unsigned char (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Amit Kumar](https://openjdk.org/census#amitkumar) (@offamitkumar - Committer)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19567/head:pull/19567` \
`$ git checkout pull/19567`

Update a local copy of the PR: \
`$ git checkout pull/19567` \
`$ git pull https://git.openjdk.org/jdk.git pull/19567/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19567`

View PR using the GUI difftool: \
`$ git pr show -t 19567`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19567.diff">https://git.openjdk.org/jdk/pull/19567.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19567#issuecomment-2152876207)

### CODE REVIEW

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp [hunk 1]
```diff
@@ -661,7 +661,7 @@ bool CgroupController::read_numerical_key_value(const char* filename, const char
   for (; line != nullptr; line = fgets(buf, buf_len, fp)) {
     char after_key = line[key_len];
     if (strncmp(line, key, key_len) == 0
-          && isspace(after_key) != 0
+          && os::is_space(after_key) != 0
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/cgroupSubsystem_linux.cpp:664]: I think this is excessive. The value is a char so cannot be a problem. The only calls to isspace that need checking are those that actually pass an int, and even then there could be adequate safeguards in place.

File: src/hotspot/os/linux/os_linux.cpp [hunk 2]
```diff
@@ -1353,7 +1353,7 @@ void os::Linux::capture_initial_stack(size_t max_size) {
       i = 0;
       if (s) {
         // Skip blank chars
-        do { s++; } while (s && isspace(*s));
+        do { s++; } while (s && os::is_space(*s));
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.cpp:1356]: Again not needed.

**Review by dholmes-ora — COMMENTED** [review]
Sorry I think this is well-intentioned but unnecessary in nearly all cases. If you pass a char there is no potential problem. Only passing an actual int could be a problem.

**Comment by tstuefe** [comment]
@dholmes-ora 

> Sorry I think this is well-intentioned but unnecessary in nearly all cases. If you pass a char there is no potential problem. Only passing an actual int could be a problem.

Note that the issue was motivated by an Oracle engineer complaining about me using isspace on a char. That caused me to look up its behavior. Recently, we seem intent on eliminating UB, so why not.

That said, I agree that we probably don't need the wrapper. And casting to int feels awkward. I propose
- input from trusted sources, e.g. proc fs, don't need casting
- input from other sources should be casted to unsigned char (see recommendation here: https://en.cppreference.com/w/cpp/string/byte/isspace "To use these functions safely with plain chars (or signed chars), the argument should first be converted to unsigned char")

**Comment by dholmes-ora** [comment]
> "To use these functions safely with plain chars (or signed chars), the argument should first be converted to unsigned char"

@tstuefe wow! Okay. That is a surprise to me. A cast to unsigned char doesn't actually do anything. Every char is "representable" as an unsigned char because it holds a bit pattern between 0x00 and 0xff i.e. the function is well defined if the incoming int is either EOF (int -1) or else in the range 0x00 to 0xff. But I did a bit of searching and it seems it comes down to potential arithmetic operations on the "char" the might behave differently depending on the signed-ness. :(

File: test/hotspot/gtest/runtime/test_os.cpp [hunk 3]
```diff
@@ -39,6 +39,8 @@
 #include "os_windows.hpp"
 #endif
 
+# include <stdio.h>
+
```
Comment (jdksjolen) [inline test/hotspot/gtest/runtime/test_os.cpp:43]: Not necessary.

File: src/hotspot/share/runtime/os.cpp [hunk 4]
```diff
@@ -93,6 +93,13 @@ os::PageSizes     os::_page_sizes;
 
 DEBUG_ONLY(bool os::_mutex_init_done = false;)
 
+int os::is_space(int c) {
```
Comment (jdksjolen) [inline src/hotspot/share/runtime/os.cpp:96]: `os::isspace`

**Review by jdksjolen — COMMENTED** [review]
Hi Robert,

Here's a third opinion:

The wrapper is fine, but it should mimic the name of the function it wraps exactly: `os::isspace`. It's also good that it does the range check as regular code as opposed to an assert, as this function is used to parse external input. It's fine to be excessive when parsing strings as those parts of the code are not really performance critical.

All the best,
Johan

**Comment by roberttoyonaga** [comment]
Thanks for the feedback everyone! 

> But I did a bit of searching and it seems it comes down to potential arithmetic operations on the "char" the might behave differently depending on the signed-ness. :(

Ok, so it seems like we ought to cast to unsigned char whenever char is passed, even if the representable range is the same. 

One more argument in favor of the current wrapper approach is that it makes it easy to at-a-glance confirm `isspace` is being used correctly everywhere.

However, I am fine with just using casts wherever there is char and removing the wrapper. This is simpler and also avoids the cast to int. But what about when an int is passed to `isspace`? Maybe having the range check here would be better than casting to unsigned char or leaving as is?

**Comment by dholmes-ora** [comment]
> But what about when an int is passed to isspace?

The current casts to int are incorrect as a negative value would be sign-extended and then fail the range check. I think we have to cast to unsigned char in all cases in the caller, which renders the wrapper and range-check obsolete AFAICS.

**Comment by roberttoyonaga** [comment]
Ok I've got rid of the wrapper and replaced it with casts to unsigned char.

**Review by dholmes-ora — APPROVED** [review]
Okay. I hate the fact we have to do this, but okay. 

Thanks

**Review by offamitkumar — APPROVED** [review]
It was a good learning; You might want to update the copyright header for few of the files. 

Thanks,

**Review by tstuefe — APPROVED** [review]
Yes, unfortunate, but ok. Thank you for doing this.

**Review by TheShermanTanker — APPROVED** [review]
Had I seen this earlier I would have suggested more C\+\+ styled casts to make which conversion exactly is happening clearer, but oh well. I agree that it's unfortunate that this has to be done

**Pull request closed without merging**